src/utils/provider-connection.ts

```diff
@@ -0,0 +1,112 @@
+import { decoding, encoding } from "lib0";
+import * as awarenessProtocol from "y-protocols/awareness";
+import * as syncProtocol from "y-protocols/sync";
+import { getYDoc } from "./ws-shared-doc";
+
+const messageSync = 0;
+const messageAwareness = 1;
+
+const pingTimeout = 30000;
+
+const providerMessageResolver = (conn, message, doc) => {
+  const encoder = encoding.createEncoder();
+  const decoder = decoding.createDecoder(message);
+  const messageType = decoding.readVarUint(decoder);
+  switch (messageType) {
+    case messageSync:
+      encoding.writeVarUint(encoder, messageSync);
+      syncProtocol.readSyncMessage(decoder, encoder, doc, null);
+      if (encoding.length(encoder) > 1) {
+        doc.send(conn, encoding.toUint8Array(encoder));
+      }
+      // TODO:
+      // doc is now the updated document
+      // we need to figure out how to persist this
+      break;
+    case messageAwareness: {
+      awarenessProtocol.applyAwarenessUpdate(
+        doc.awareness,
+        decoding.readVarUint8Array(decoder),
+        conn
+      );
+      break;
+    }
+  }
+};
+
+/**
+ * Setup a new client
+ * @param {any} conn
+ * @param {any} req
+ * @param {any} options
+ */
+const setupProviderConnection = (
+  conn,
+  req,
+  { docName = req.url.slice(1).split("/")[1], gc = true } = {}
```

Comment:
I think there are a couple different refactors that are happening here, but its hard to see what they are because we are moving things around at the same time. It's better to keep these refactors on different commits so I can see whats changing.

On the other hand, its bad form to do refactors on things you just added. Let's squash these commits together and try to find a different way to break them apart.

Try to think about making your changes in steps:
1. [refactor] Move the signaling service to `/signal`
2. [feat] Add the provider servicer to `/provider`

Reply:
I'll get on that - would it be good to put the change of `{ docName = req.url.slice(1).split("?")[1], gc = true } = {}` to `{ docName = req.url.slice(1).split("/")[1], gc = true } = {}` on a separate [fix] commit? I don't think splitting on "?" is correct because there isn't that character in the url, so I changed it to "/" to get the room name. This kind of change warrants it's own commit I guess is what I'm asking?